pyuvdata/uvcal/uvcal.py

```diff
@@ -3351,6 +3355,446 @@ def _convert_to_filetype(self, filetype):
+                        uvc.freq_range = np.squeeze(freq_range_use).tolist()
+                    else:
+                        raise ValueError(
+                            "if future_array_shapes is False, freq_range must have "
+                            "2 elements."
+                        )
+
+        for param_name in params_to_copy:
+            setattr(uvc, param_name, getattr(uvdata, param_name))
+
+        # sort the antenna information (the order in the UVData object may be strange)
+        ant_order = np.argsort(uvc.antenna_numbers)
+        uvc.antenna_numbers = uvc.antenna_numbers[ant_order]
+        uvc.antenna_names = ((np.asarray(uvc.antenna_names))[ant_order]).tolist()
+        uvc.antenna_positions = uvc.antenna_positions[ant_order, :]
+
+        if times is None:
+            # get all unique times
+            uvc.time_array = np.unique(uvdata.time_array)
+            uvdata_int_times = np.unique(uvdata.integration_time)
+            if uvdata_int_times.size == 1:
+                uvdata_int_times = uvdata_int_times[0]
+                if uvc.future_array_shapes:
+                    uvc.integration_time = np.full(
+                        uvc.time_array.size, uvdata_int_times, dtype=np.float64
+                    )
+                else:
+                    uvc.integration_time = uvdata_int_times
+            else:
+                raise ValueError(
+                    "uvdata integration times vary. Please specify times and "
+                    "integration_time"
+                )
+        else:
+            uvc.time_array = times
+            if integration_time is None:
+                raise ValueError(
+                    "integation_time must be provided if times is provided"
+                )
+            if future_array_shapes:
+                if isinstance(integration_time, (np.ndarray, list)):
+                    uvc.integration_time = np.asarray(integration_time)
+                else:
+                    uvc.integration_time = np.full(
+                        uvc.time_array.size, integration_time, dtype=np.float64
+                    )
+            else:
+                if isinstance(integration_time, (np.ndarray, list)):
+                    raise ValueError(
+                        "integration_time must be scalar if future_array_shapes is "
+                        "False."
+                    )
+                uvc.integration_time = integration_time
+        uvc.Ntimes = uvc.time_array.size
+        uvc.set_lsts_from_time_array()
+
+        if time_range is not None:
+            uvc.time_range = time_range
+
+        if jones is None:
+            if np.all(uvdata.polarization_array < -4):
+                if uvdata.Npols == 1 and uvdata.polarization_array[0] > -7:
+                    # single pol data, make a single pol cal object
+                    uvc.jones_array = uvdata.polarization_array
+                else:
+                    uvc.jones_array = np.array([-5, -6])
```

Comment:
Very clever! One small question: do we have to worry about the order of the `jones_array`? I forget if we use it to index anything...

Reply:
We only use it to index the data-like arrays. So when the user fills out those arrays they need to line them up with the `jones_array`, but I don't think we need to worry about anything in this method.

Reply:
Sounds good, thanks for clarifying.